
# Import the dataset



In [ ]:
try:
  import google.colab
  IN_COLAB = True
except:
  IN_COLAB = False

import os

if IN_COLAB:
    if (os.getcwd() != '/content'):
        %cd ../../

    !git clone https://github.com/JeffSackmann/tennis_atp
%cd /content/JeffSackmann/tennis_atp

Cloning into 'tennis_atp'...
remote: Enumerating objects: 2399, done.
remote: Counting objects: 100% (7/7), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 2399 (delta 0), reused 1 (delta 0), pack-reused 2392
Receiving objects: 100% (2399/2399), 244.35 MiB | 18.67 MiB/s, done.
Resolving deltas: 100% (1784/1784), done.
Checking out files: 100% (174/174), done.
[Errno 2] No such file or directory: '/content/JeffSackmann/tennis_atp'
/content


# Import libraries

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
%cd /content/tennis_atp/
ext = '.csv'
#full data from 1991
start1 = 1991 #1968
end1 = 2021
ft1 = 'atp_matches_'

start2 = 1991
end2 = 2021
ft2 = 'atp_matches_futures_'


start3 = 1991
end3 = 2021
ft3 = 'atp_matches_qual_chall_'

players = 'atp_players.csv'

rankings = ['atp_rankings_20s.csv','atp_rankings_00s.csv','atp_rankings_10s.csv','atp_rankings_70s.csv','atp_rankings_80s.csv','atp_rankings_90s.csv',
            'atp_rankings_current.csv']

/content/tennis_atp


In [ ]:
def readfiles(ft,st,en):
    l = []
    for i in range(st,en+1):
        p = ft + str(i)+ ext
        print(p)
        df = pd.read_csv(p)
        l.append(df)
        #print(df.head(2))
    return l

Read all the matches files.

In [ ]:
f1 = readfiles(ft1,start1,end1)

atp_matches_1991.csv
atp_matches_1992.csv
atp_matches_1993.csv
atp_matches_1994.csv
atp_matches_1995.csv
atp_matches_1996.csv
atp_matches_1997.csv
atp_matches_1998.csv
atp_matches_1999.csv
atp_matches_2000.csv
atp_matches_2001.csv
atp_matches_2002.csv
atp_matches_2003.csv
atp_matches_2004.csv
atp_matches_2005.csv
atp_matches_2006.csv
atp_matches_2007.csv
atp_matches_2008.csv
atp_matches_2009.csv
atp_matches_2010.csv
atp_matches_2011.csv
atp_matches_2012.csv
atp_matches_2013.csv
atp_matches_2014.csv
atp_matches_2015.csv
atp_matches_2016.csv
atp_matches_2017.csv
atp_matches_2018.csv
atp_matches_2019.csv
atp_matches_2020.csv
atp_matches_2021.csv


Read all the ranking files.

In [ ]:
ranks = []
for i in rankings:
    print(i)
    ranks.append(pd.read_csv(i))

atp_rankings_20s.csv
atp_rankings_00s.csv
atp_rankings_10s.csv
atp_rankings_70s.csv
atp_rankings_80s.csv
atp_rankings_90s.csv
atp_rankings_current.csv


In [ ]:
all_matches = pd.concat(f1)

In [ ]:
print(all_matches.head())
cat_cols = ['tourney_id','tourney_name','surface','draw_size','tourney_level','tourney_date','winner_id']

  tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0   1991-339     Adelaide    Hard         32             A      19901231   
1   1991-339     Adelaide    Hard         32             A      19901231   
2   1991-339     Adelaide    Hard         32             A      19901231   
3   1991-339     Adelaide    Hard         32             A      19901231   
4   1991-339     Adelaide    Hard         32             A      19901231   

   match_num  winner_id  winner_seed winner_entry  ... l_1stIn l_1stWon  \
0          1     101723          NaN          NaN  ...    62.0     44.0   
1          2     100946          NaN            Q  ...    41.0     35.0   
2          3     101234          NaN          NaN  ...    37.0     22.0   
3          4     101889          8.0          NaN  ...    45.0     30.0   
4          5     101274          NaN          NaN  ...    41.0     28.0   

   l_2ndWon l_SvGms  l_bpSaved  l_bpFaced  winner_rank winner_rank_points  \
0      23.0    

In [ ]:
all_matches.shape


(99360, 49)

In [ ]:
all_matches.describe()

draw_size  tourney_date     match_num      winner_id   winner_seed  \
count  99360.000000  9.936000e+04  99360.000000   99360.000000  40170.000000   
mean      53.007568  2.004922e+07     63.579006  104943.220018      6.873214   
std       38.242706  8.848977e+04    107.630448   10695.233105      6.164352   
min        2.000000  1.990123e+07      1.000000  100284.000000      1.000000   
25%       32.000000  1.997062e+07      9.000000  102053.000000      3.000000   
50%       32.000000  2.004102e+07     23.000000  103454.000000      5.000000   
75%       64.000000  2.012092e+07     53.000000  104655.000000      8.000000   
max      128.000000  2.021120e+07   1701.000000  210107.000000     35.000000   

          winner_ht    winner_age       loser_id    loser_seed      loser_ht  \
count  95879.000000  99311.000000   99360.000000  21977.000000  92590.000000   
mean     185.622671     25.772929  105141.185789      8.223370    185.139550   
std        6.990696      3.764075   11290.482067      6.653615      6.799341   
min        1.700000     14.351814  100282.000000      1.000000      1.700000   
25%      180.000000     22.970568  102106.000000      4.000000    180.000000   
50%      185.000000     25.503080  103428.000000      6.000000    185.000000   
75%      190.000000     28.287474  104678.000000     11.000000    190.000000   
max      211.000000     42.792608  210274.000000     35.000000    211.000000   

       ...       l_1stIn      l_1stWon      l_2ndWon       l_SvGms  \
count  ...  89475.000000  89475.000000  89475.000000  89476.000000   
mean   ...     47.919877     31.827103     15.002626     12.188710   
std    ...     19.388608     14.449350      7.228977      4.127567   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...     34.000000     22.000000     10.000000      9.000000   
50%    ...     44.000000     29.000000     14.000000     11.000000   
75%    ...     58.000000     39.000000     19.000000     15.000000   
max    ...    328.000000    284.000000    101.000000     91.000000   

          l_bpSaved     l_bpFaced   winner_rank  winner_rank_points  \
count  89475.000000  89475.000000  98315.000000        97323.000000   
mean       4.813434      8.750768     81.048589         1410.054591   
std        3.274505      4.130267    135.134479         1749.619712   
min       -6.000000      0.000000      1.000000            1.000000   
25%        2.000000      6.000000     18.000000          522.000000   
50%        4.000000      8.000000     46.000000          871.000000   
75%        7.000000     11.000000     89.000000         1577.000000   
max       28.000000     38.000000   2101.000000        16950.000000   

         loser_rank  loser_rank_points  
count  97045.000000       96056.000000  
mean     119.439518         882.222474  
std      180.018822         989.287834  
min        1.000000           1.000000  
25%       37.000000         390.000000  
50%       71.000000         649.000000  
75%      119.000000        1030.000000  
max     2159.000000       16950.000000  

[8 rows x 35 columns]

In [ ]:
all_matches.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 99360 entries, 0 to 2726
Data columns (total 49 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tourney_id          99360 non-null  object 
 1   tourney_name        99360 non-null  object 
 2   surface             99360 non-null  object 
 3   draw_size           99360 non-null  int64  
 4   tourney_level       99360 non-null  object 
 5   tourney_date        99360 non-null  int64  
 6   match_num           99360 non-null  int64  
 7   winner_id           99360 non-null  int64  
 8   winner_seed         40170 non-null  float64
 9   winner_entry        11993 non-null  object 
 10  winner_name         99360 non-null  object 
 11  winner_hand         99351 non-null  object 
 12  winner_ht           95879 non-null  float64
 13  winner_ioc          99360 non-null  object 
 14  winner_age          99311 non-null  float64
 15  loser_id            99360 non-null  int64  
 16  loser

In [ ]:
length = len(all_matches)/2
print(length)
y= []
for i in range(int(length)):
    y.append(0)
    
for i in range(int(length)):
    y.append(1)
    
all_matches['y'] = pd.Series(y, index=all_matches.index)

49680.0


Aggiungo la colonna y a all_matches.

La prima parte del dataset (le prime 49680 partite) avranno y=0, e la restante parte avrà y=1

In [ ]:
all_matches1 = all_matches[:int(length)]
all_matches1.to_csv('all_matches_firsthalf.csv', encoding='utf-8', index=False)
#print(all_matches1.shape)
#print(all_matches1.tail())
all_matches1.head()

tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0   1991-339     Adelaide    Hard         32             A      19901231   
1   1991-339     Adelaide    Hard         32             A      19901231   
2   1991-339     Adelaide    Hard         32             A      19901231   
3   1991-339     Adelaide    Hard         32             A      19901231   
4   1991-339     Adelaide    Hard         32             A      19901231   

   match_num  winner_id  winner_seed winner_entry  ... l_1stWon l_2ndWon  \
0          1     101723          NaN          NaN  ...     44.0     23.0   
1          2     100946          NaN            Q  ...     35.0     27.0   
2          3     101234          NaN          NaN  ...     22.0      6.0   
3          4     101889          8.0          NaN  ...     30.0     11.0   
4          5     101274          NaN          NaN  ...     28.0     15.0   

   l_SvGms l_bpSaved  l_bpFaced  winner_rank  winner_rank_points loser_rank  \
0     16.0       6.0        8.0         56.0                 NaN        2.0   
1     15.0       1.0        2.0        304.0                 NaN       75.0   
2      8.0       4.0        8.0         82.0                 NaN       69.0   
3     10.0       5.0        8.0         50.0                 NaN       84.0   
4     11.0       4.0        8.0         88.0                 NaN       28.0   

  loser_rank_points  y  
0               NaN  0  
1               NaN  0  
2               NaN  0  
3               NaN  0  
4               NaN  0  

[5 rows x 50 columns]

In [ ]:
all_matches2 = all_matches[int(length):]
all_matches2.to_csv('all_matches_secondhalf.csv', encoding='utf-8', index=False)
all_matches2.head()

tourney_id                 tourney_name surface  draw_size tourney_level  \
3098  2004-D036  Davis Cup G1 R2: JPN vs IND    Hard          4             D   
3099  2004-D036  Davis Cup G1 R2: JPN vs IND    Hard          4             D   
3100  2004-D036  Davis Cup G1 R2: JPN vs IND    Hard          4             D   
3101  2004-D036  Davis Cup G1 R2: JPN vs IND    Hard          4             D   
3102  2004-D037  Davis Cup G1 PO: TPE vs PAK    Hard          4             D   

      tourney_date  match_num  winner_id  winner_seed winner_entry  ...  \
3098      20040409          1     102300          NaN          NaN  ...   
3099      20040409          2     102233          NaN          NaN  ...   
3100      20040409          4     102863          NaN          NaN  ...   
3101      20040409          5     102300          NaN          NaN  ...   
3102      20040409          1     104229          NaN          NaN  ...   

     l_1stWon l_2ndWon  l_SvGms l_bpSaved  l_bpFaced  winner_rank  \
3098      NaN      NaN      NaN       NaN        NaN        275.0   
3099      NaN      NaN      NaN       NaN        NaN          NaN   
3100      NaN      NaN      NaN       NaN        NaN        206.0   
3101      NaN      NaN      NaN       NaN        NaN        275.0   
3102      NaN      NaN      NaN       NaN        NaN        110.0   

      winner_rank_points loser_rank loser_rank_points  y  
3098               126.0      266.0             133.0  1  
3099                 NaN      206.0             177.0  1  
3100               177.0      266.0             133.0  1  
3101               126.0        NaN               NaN  1  
3102               374.0        NaN               NaN  1  

[5 rows x 50 columns]

In [ ]:
L  = list(all_matches1)
for i in range(17,27):
    L[i],L[i-10]=L[i-10],L[i]
all_matches1=all_matches1.reindex(columns=L)

In all_matches1 ho y=0, modifico le colonne in modo tale che loser=Player1 e winner=Player2. Quindi y=0 significa che il vincitore è il Player2.

Al contrario in all_matches2, dove y=1, quindi il vincitore è il Player1 e rinomino le colonne.

In [ ]:
all_matches1.rename(columns={'loser_id':'Player1_id', 'loser_seed':'Player1_seed', 'loser_entry':'Player1_entry', 'loser_name':'Player1_name', 'loser_hand':'Player1_hand', 'loser_ht':'Player1_ht', 'loser_ioc':'Player1_ioc', 'loser_age':'Player1_age', 'loser_rank':'Player1_rank', 'loser_rank_points':'Player1_rank_points','winner_id':'Player2_id', 'winner_seed':'Player2_seed', 'winner_entry':'Player2_entry', 'winner_name':'Player2_name', 'winner_hand':'Player2_hand', 'winner_ht':'Player2_ht', 'winner_ioc':'Player2_ioc', 'winner_age':'Player2_age', 'winner_rank':'Player2_rank', 'winner_rank_points':'Player2_rank_points'},inplace=True)


In [ ]:
all_matches2.rename(columns={'loser_id':'Player2_id', 'loser_seed':'Player2_seed', 'loser_entry':'Player2_entry', 'loser_name':'Player2_name', 'loser_hand':'Player2_hand', 'loser_ht':'Player2_ht', 'loser_ioc':'Player2_ioc', 'loser_age':'Player2_age', 'loser_rank':'Player2_rank', 'loser_rank_points':'Player2_rank_points','winner_id':'Player1_id', 'winner_seed':'Player1_seed', 'winner_entry':'Player1_entry', 'winner_name':'Player1_name', 'winner_hand':'Player1_hand', 'winner_ht':'Player1_ht', 'winner_ioc':'Player1_ioc', 'winner_age':'Player1_age', 'winner_rank':'Player1_rank', 'winner_rank_points':'Player1_rank_points'},inplace=True)


/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:5047: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [ ]:
cc = [all_matches1,all_matches2]
all_matches_final= pd.concat(cc)

In [ ]:
from sklearn.utils import shuffle
all_matches_final = shuffle(all_matches_final)
all_matches_final.to_csv('all_matches_final.csv', encoding='utf-8', index=False)

In [ ]:
all_matches_final

tourney_id                 tourney_name surface  draw_size tourney_level  \
3033   1993-560                      US Open    Hard        128             G   
2515   1993-421               Canada Masters    Hard         56             M   
2802   2015-337                       Vienna    Hard         32             A   
1791   2003-468                      Estoril    Clay         32             A   
1719   2002-441                  Long Island    Hard         48             A   
...         ...                          ...     ...        ...           ...   
1902  2015-D010  Davis Cup WG QF: AUS vs KAZ   Grass          4             D   
1242   2014-308                       Munich    Clay         28             A   
880    2000-403                Miami Masters    Hard         96             M   
1089   2008-414              Hamburg Masters    Clay         56             M   
1735   2011-540                    Wimbledon   Grass        128             G   

      tourney_date  match_num Player1_entry            Player1_name  \
3033      19930830         55             Q  Jean Philippe Fleurian   
2515      19930726         21           NaN             Wally Masur   
2802      20151019         18           NaN           Fabio Fognini   
1791      20030407         12           NaN          Sjeng Schalken   
1719      20020819         25           NaN           Adrian Voinea   
...            ...        ...           ...                     ...   
1902      20150717          1           NaN       Mikhail Kukushkin   
1242      20140428          4           NaN           Jurgen Melzer   
880       20000320         25             Q        David Nalbandian   
1089      20080511          6            LL               Ivo Minar   
1735      20110620         94           NaN          Michael Llodra   

     Player1_hand  ...  l_1stWon l_2ndWon  l_SvGms l_bpSaved  l_bpFaced  \
3033            R  ...      43.0     24.0     17.0       6.0       11.0   
2515            R  ...      33.0     19.0     14.0       2.0        5.0   
2802            R  ...      19.0     12.0      8.0       4.0        7.0   
1791            R  ...      22.0      4.0      8.0       7.0       12.0   
1719            R  ...      45.0     23.0     16.0       3.0        4.0   
...           ...  ...       ...      ...      ...       ...        ...   
1902            R  ...       NaN      NaN      NaN       NaN        NaN   
1242            L  ...       9.0      8.0      7.0       2.0        7.0   
880             R  ...      38.0     20.0     15.0       6.0       10.0   
1089            R  ...      30.0     11.0      9.0       6.0        8.0   
1735            L  ...      56.0     17.0     17.0      10.0       15.0   

     Player2_rank  Player2_rank_points  Player1_rank  Player1_rank_points  y  
3033        109.0                397.0         125.0                341.0  0  
2515         46.0                682.0          23.0               1160.0  0  
2802        259.0                187.0          22.0               1470.0  1  
1791         61.0                653.0          14.0               1700.0  0  
1719         18.0               1415.0          63.0                619.0  0  
...           ...                  ...           ...                  ... ..  
1902         69.0                718.0          63.0                747.0  1  
1242        765.0                 25.0          66.0                700.0  1  
880          60.0                697.0         473.0                 49.0  0  
1089         32.0                955.0          70.0                538.0  1  
1735         89.0                574.0          35.0               1195.0  1  

[99360 rows x 50 columns]